src/urllib3/response.py

```diff
@@ -701,14 +702,46 @@ def read(
         flush_decoder = False
         fp_closed = getattr(self._fp, "closed", False)
 
+        # Reading more than 2 GiB (`int` max value in C) via SSL on some
+        # CPython 3.8, 3.9, and 3.10 versions leads to an overflow error
+        # that has to be prevented if `amt` or `self.length_remaining`
+        # indicate that it may happen.
+        c_int_max = (2**31) - 1
+        if (
+            (amt and amt > c_int_max)
+            or (self.length_remaining and self.length_remaining > c_int_max)
+        ) and (
+            (3, 8, 0, "a", 4) <= sys.version_info < (3, 9, 7)
+            or (3, 10, 0, "b", 1) <= sys.version_info < (3, 10, 0, "rc", 1)
```

Comment:
Let's not consider unreleased versions of Python please :)

Reply:
Relocating a comment from https://github.com/illia-v/urllib3/commit/2d6cf6c6f2714e5ad6455c1e665af83e9af9293f#r76646438 (thank you @illia-v for redirecting here):

Would it be worth considering using [`util.ssl_._is_ge_openssl_v1_1_1`](https://github.com/urllib3/urllib3/blob/f9c7434090ce96ca93180137356531a9cfcce822/src/urllib3/util/ssl_.py#L28-L38) -- or something like it -- to check whether to enable this logic?

One of my thoughts was that that could allow people who compile `cpython` themselves (for example, with different SSL libraries/version) and who use other variants of Python to benefit from the fix.

Reply:
Sure, if this fix depends on the OpenSSL version, we should definitely check that instead. I haven't investigated enough to tell.

Reply:
@jayaddison I can reproduce the overflow on CPython 3.8 compiled against OpenSSL 1.0.2 and 1.1.1. Does it really depend on the OpenSSL version?

Reply:
I was also under the impression that this was an issue in Python, not with OpenSSL.

Reply:
Hmm, yep, I think I misunderstood something.  Re-constructing the situation:

The underlying 2GB limitation derives from OpenSSL <1.1.1, which used signed 32-bit ints ([bpo-42853 note](https://bugs.python.org/issue42853#msg384566)) in the relevant library function (`SSL_read`).

However, Python (cpython?) from 3.10 onwards calls the [more-recent](https://github.com/openssl/openssl/commit/4ee7d3f94590d3766c28eed03b969f338ddf96dc) [`SSL_read_ex`](https://github.com/python/cpython/blob/3.10/Modules/_ssl.c#L2513) function in OpenSSL that _does_ provide >2GB support.

So: an old Python compiled with OpenSSL >=1.1.1 would still have the limitation/bug, and Python with OpenSSL <1.1.1 would fail to compile in the first place (because the `SSL_read_ex` function wouldn't exist to link to).

Please double-check my logic there - but yep, I think I was mistaken, sorry for the distraction.

Reply:
@jayaddison No problem, thanks for participating in the discovery. It's good to know the innards of the space we're working on! :)

Reply:
> Let's not consider unreleased versions of Python please :)

OK, I updated the condition in 49ed9a1

Reply:
BTW, if you know how to check whether a response is read via SSL, please let me know.